Replacing creation of archive file with testing Zipf's law for the books

lexnederbragt opened the pull request:
This PR changes the final step of the 'pipeline'. Instead of creating an archive file, Zipf's law is tested for all books. A new script, `zipf_test.py`, creates a summary table file, `results.txt`, with the counts of the two most frequent words and their ratio:

```
Book	First	Second	Ratio
abyss	4044	2807	1.44
isles	3822	2460	1.55
last	12244	5566	2.20
sierra	4242	2469	1.72
```

The motivation for the change is to add a final step that creates a 'useful' result, which is updated when new books are added - or when one of the books is 'updated'. This reflects a 'real' scenario with intermediate steps and a final collecting step.

Some additional notes:

* This change adds an interesting new illustration of the power of `make` at the end of `04-dependencies`: adding `wordcloud.py` as dependency to the `.dat` files; as a result, `results.txt` will be recreated as well *without* adding any new dependencies.
* I added a new challenge to `04-dependencies`, testing the effect of updating only one of the input files (one of the books). It demonstrates how only the relevant parts of the 'pipeline' get updated.
* For the final challenge, the creation of the PNG files needs to be added as a dependency of the `results.txt` file, otherwise they are not (re)created. This feels a bit artificial but I couldn't come up with any other solution...
* Commit [#6a92e48](https://github.com/swcarpentry/make-novice/pull/43/commits/6a92e4849261d0be64e0d93eec2329dd430d2606) removes the need for manually cropping the `img/07-functions.png` and `img/08-conclusion-challenge.png` images to remove the 'Makefile' node. The README file is updated to reflect this.

mikej888 commented on 01-intro.md:
```diff
@@ -194,8 +209,8 @@ Another popular option is to comment out a subset of the lines in
 
-# Archive the results so we can share them with a colleague.
-tar -czf analysis.tar.gz isles.dat abyss.dat
+# This lines is also commented out because it doesn't need to be rerun.
```
Typo: "This line is"

mikej888 commented on 03-variables.md:
```diff
@@ -53,20 +53,20 @@ Makefile but forget to rename it elsewhere.
 
-In our `analysis.tar.gz` rule we duplicate the data file names and the
+In our `results.txt` rule we duplicate the data file names and the
 archive name:
```
"archive name" => "results file name"

mikej888 commented on 03-variables.md:
```diff
@@ -53,20 +53,20 @@ Makefile but forget to rename it elsewhere.
+        python zipf_test.py abyss.dat isles.dat last.dat > results.txt
 ~~~
 
 Looking at the archive name first, we can replace it in the action
```
"archive name" => "results file name"

mikej888 commented on 03-variables.md:
```diff
@@ -145,16 +145,16 @@ $ make analysis.tar.gz
-rule'. This works well for `analysis.tar.gz` as its action 
+rule'. This works well for `results.txt` as its action
 treats all the dependencies the same - as the contents of the
 archive.
```
"archive " => "results file"

mikej888 commented on 04-dependencies.md:
```diff
@@ -78,31 +78,57 @@ python wordcount.py books/abyss.txt abyss.dat
+> 4. all `.dat` and `results.txt` are recreated
+
+
+> ## `workcloud.py` as dependency of `results.txt` {.challenge}
```
"workcloud.py" => "wordcount.py"

mikej888 commented on 04-dependencies.md:
```diff
@@ -78,31 +78,57 @@ python wordcount.py books/abyss.txt abyss.dat
+
+> ## `workcloud.py` as dependency of `results.txt` {.challenge}
+>
+> What would happen if you actually added `workcloud.py` as dependency of `results.txt`, and why?
```
"workcloud.py" => "wordcount.py"

mikej888 commented on 06-variables.md:
```diff
@@ -52,8 +52,9 @@ used to implement our script from Python to R).
 >
-> Update `Makefile` so that the `%.dat` and `analysis.tar.gz` rules
+> Update `Makefile` so that the `%.dat` rule
 > reference the variables `COUNT_SRC` and `COUNT_EXE`.
```
"reference" => "references"

mikej888 commented on solutions.md:
```diff
@@ -111,42 +110,88 @@ isles.dat : books/isles.txt
+'Follow' the decency tree to understand the answer(s).
+
+> ## `workcloud.py` as dependency of `results.txt` {.challenge}
+>
```
"workcloud.py" => "wordcount.py"

mikej888 commented on solutions.md:
```diff
@@ -111,42 +110,88 @@ isles.dat : books/isles.txt
+>
+> What would happen if you actually added `workcloud.py` as dependency of `results.txt`, and why?
+
+`workcloud.py` becomes a part of `$^`, thus the command becomes
```
"workcloud.py" => "wordcount.py"

mikej888 commented on solutions.md:
```diff
@@ -111,42 +110,88 @@ isles.dat : books/isles.txt
+`workcloud.py` becomes a part of `$^`, thus the command becomes
+
+~~~ {.bash}
+python zipf_test.py abyss.dat isles.dat last.dat workcloud.py > results.txt
```
"workcloud.py" => "wordcount.py"

mikej888 commented on solutions.md:
```diff
@@ -111,42 +110,88 @@ isles.dat : books/isles.txt
+    counts = load_word_counts(input_file)
+  File "path/to/wordcount.py", line 35, in load_word_counts
+    with open(filename, "r") as input_fd:
+IOError: [Errno 2] No such file or directory: 'workcloud.py'
```
"workcloud.py" => "wordcount.py"

mikej888 commented on 08-conclusion.md:
```diff
@@ -28,12 +28,12 @@ papers.
 > 
 > * Create `.png` files from `.dat` files using `plotcount.py`.
-> * Add the script and `.png` files to the archive.
+> * Add the script and `.png` files as dependencies of the `results.txt` file to make sure they are (re)created when needed.
```
It feels wrong to add the `.png` files as dependent on `results.txt` as this is artificial. It'd be cleaner to just extend `all` to build both `results.txt` and the `.png` files.

mikej888 commented:
Thanks for the pull request. It looks good. My only real issue is with:

> For the final challenge, the creation of the PNG files needs to be added as a dependency of the results.txt file, otherwise they are not (re)created. This feels a bit artificial but I couldn't come up with any other solution...

I'd rather the `.png` files be added to `all` as a dependency, rather than introducing a non-existent dependency (and teaching attendees a bad practice).

lexnederbragt commented:
Thanks - and sorry for the typo's :-). I agree with your solution for the `.png` files and the `all` target. Of course!

This means however, that the final `make` command can no longer be `make results.txt` as it would then skip the `.png` files. I addressed this in the challenge in 08-conclusion (and adjusted the lessons' Makefile accordingly). The final dependency graph looks very nice now!

mikej888 commented:
Thanks @lexnederbragt! Yes, the final dependency graph looks far richer now!
